examples/flutter_gallery/android/app/src/main/java/io/flutter/examples/gallery/FlutterGalleryInstrumentation.java

```diff
@@ -0,0 +1,40 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.examples.gallery;
+
+import io.flutter.plugin.common.MethodCall;
+import io.flutter.plugin.common.MethodChannel;
+import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
+import io.flutter.plugin.common.MethodChannel.Result;
+import io.flutter.view.FlutterView;
+
+/** Instrumentation for testing using Android Espresso framework. */
+public class FlutterGalleryInstrumentation implements MethodCallHandler {
+
+  private boolean testFinished;
+  private boolean testSuccessful;
+
+  FlutterGalleryInstrumentation(FlutterView view) {
+    new MethodChannel(view, "io.flutter.examples.gallery/TestLifecycleListener")
+        .setMethodCallHandler(this);
+  }
+
+  @Override
+  public void onMethodCall(MethodCall call, Result result) {
+    testSuccessful = call.method.equals("success");
+    testFinished = true;
+    result.success(null);
+  }
+
+  public boolean isTestSuccessful() {
+    return testSuccessful;
+  }
+
+  public void waitForTestToFinish() throws Exception {
+    while(!testFinished) {
+      Thread.sleep(200);
```

Comment:
You can avoid polling with a synchronization object such as android.os.ConditionVariable

If you need to poll, then mark the testFinished flag as volatile